### PR TITLE
Add a merge policy that prunes ID postings for soft-deleted but retained documents

### PR DESCRIPTION
**Pull request by s1monw**
This change adds a merge policy that drops all `_id` postings for documents that
are marked as soft-deleted but retained across merges. This is usually unnecessary unless soft-deletes are used with a retention policy since otherwise a merge would remove deleted documents anyway.

Yet, this merge policy prevents extreme cases where a very large number of soft-deleted
documents are retained and are impacting update performance.
Note, using this merge policy will remove all lookup by ID capabilities for soft-deleted documents.

### CODE REVIEW

**Comment by jpountz** [comment]
I'm undecided about whether this is the right trade-off in general given that this change would mean that the inverted index and doc values would no longer be consistent, which isn't something that we rely on today, but that we have thought about relying on in the past at a couple occasions. Maybe we will get there eventually but what about only doing this on the `_id` field for now? While this won't help with search, it should already help significantly update performance in worst-case scenarios?

**Comment by ywelsch** [comment]
> Maybe we will get there eventually but what about only doing this on the _id field for now?

+1 to exploring & benchmarking this. While this should help with index/update performance, I wonder mainly whether merge performance of other indexed fields will still be hampered by a large number of soft-deleted docs.

**Comment by s1monw** [comment]
@jpountz @ywelsch @dnhatn I did benchmark this and didn't see any slowdowns on our standard benchmarks. When @dliappis benchmarked it we were actually able to run the benchmark with this change due to the high retention the benchmark failed quickly with this change. Now compared to the initial version this change only prunes the `_id` field for deleted docs to guarantee update performance. I would appreciate a review. Compared to the previous versions I also added code that removes terms without any docs.

File: server/src/main/java/org/elasticsearch/index/engine/PrunePostingsMergePolicy.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                                                OnlyLiveDocsPostingsEnum reuseInstance = (OnlyLiveDocsPostingsEnum) reuse;
+                                                reuseInstance.reset(super.postings(reuseInstance.in, flags));
+                                                return reuseInstance;
+                                            }
+                                            return new OnlyLiveDocsPostingsEnum(super.postings(null, flags), liveDocs);
+                                        }
+
+                                        @Override
+                                        public ImpactsEnum impacts(int flags) throws IOException {
+                                            throw new UnsupportedEncodingException();
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/engine/PrunePostingsMergePolicy.java]: UnsupportedOperationException?

**Review by jpountz — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/engine/PrunePostingsMergePolicy.java (hunk 1 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/index/engine/PrunePostingsMergePolicy.java]: LOL yeah

**Review by dnhatn — APPROVED** [review]
LGTM. Can you add a test with our InternalEngine?

File: server/src/main/java/org/elasticsearch/index/engine/PrunePostingsMergePolicy.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                                            internal = postings(internal, PostingsEnum.NONE);
+                                            if (internal.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                                                return AcceptStatus.YES;
+                                            }
+                                            return AcceptStatus.NO;
+                                        }
+
+                                        @Override
+                                        public PostingsEnum postings(PostingsEnum reuse, int flags) throws IOException {
+                                            if (reuse != null && reuse instanceof OnlyLiveDocsPostingsEnum) {
```
[the remark this answers is not in the document]
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/index/engine/PrunePostingsMergePolicy.java]: sure

**Comment by s1monw** [comment]
> Can you add a test with our InternalEngine?

sure  - any idea what I should test here?

**Comment by dnhatn** [comment]
> > Can you add a test with our InternalEngine?

> sure  - any idea what I should test here?

`testLookupVersionWithPrunedAwayIds` that you added is good enough for us already. Can we use InternalEngine instead of Lucene stuff in that test.

**Comment by s1monw** [comment]
> testLookupVersionWithPrunedAwayIds that you added is good enough for us already. Can we use InternalEngine instead of Lucene stuff in that test.

it's terribly bloated - tell me what you think I should assert and I will add one

**Comment by dnhatn** [comment]
> it's terribly bloated - tell me what you think I should assert and I will add one

Just make sure that InternalEngine prunes away _id of deleted documents like what you did in testLookupVersionWithPrunedAwayIds. I think we're good here.

**Comment by jasontedor** [comment]
> Now compared to the initial version this change only prunes the `_id` field for deleted docs to guarantee update performance.

I am good with this tradeoff. I would not expect future consumers of the changes API to want to filter on the `_id` field, and for users that have this use case, they can index a separate field anyway. Compared to the previous implementation the tradeoff was too far for me towards limiting future functionality, but I am +1 now. Thank you for taking another iteration.

**Review by ywelsch — APPROVED** [review]
LGTM. Please update title and description. Thank you

**Comment by s1monw** [comment]
@dnhatn  I pushed a new commit with a test

**Comment by s1monw** [comment]
> LGTM. Please update title and description. Thank you

@ywelsch done

**Review by dnhatn — APPROVED** [review]
> @dnhatn I pushed a new commit with a test

LGTM. Thank you for the extra iteration :).

**Pull request merged**